pkg/builtin/unary/ltrim.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2022 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unary
+
+import (
+	"fmt"
+
+	"github.com/matrixorigin/matrixone/pkg/builtin"
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/sql/colexec/extend"
+	"github.com/matrixorigin/matrixone/pkg/sql/colexec/extend/overload"
+	"github.com/matrixorigin/matrixone/pkg/vectorize/ltrim"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+)
+
+func init() {
+	extend.FunctionRegistry["ltrim"] = builtin.Ltrim
+
+	// register function args and returns
+	overload.AppendFunctionRets(builtin.Ltrim, []types.T{types.T_char}, types.T_char)
+	overload.AppendFunctionRets(builtin.Ltrim, []types.T{types.T_varchar}, types.T_varchar)
+
+	extend.UnaryReturnTypes[builtin.Ltrim] = func(extend extend.Extend) types.T {
+		return getUnaryReturnType(builtin.Ltrim, extend) // register a get return type function
+	}
+
+	extend.UnaryStrings[builtin.Ltrim] = func(extend extend.Extend) string {
+		return fmt.Sprintf("ltrim(%s)", extend) // define a stringify function
+	}
+
+	overload.OpTypes[builtin.Ltrim] = overload.Unary // register function type
+
+	overload.UnaryOps[builtin.Ltrim] = []*overload.UnaryOp{
+		{
+			Typ:        types.T_varchar,
+			ReturnType: types.T_varchar,
+			Fn: func(origVec *vector.Vector, proc *process.Process, _ bool) (*vector.Vector, error) {
+				origVecCol := origVec.Col.(*types.Bytes)
+
+				// totalCount - spaceCount is the total bytes need for the ltrim-ed string
+				spaceCount := ltrim.CountSpacesFromLeft(origVecCol)
+				totalCount := int32(len(origVecCol.Data))
```

Comment:
perhaps we could reuse the original vector when possible

Reply:
huh, if we reuse the original vector, it kind of like using double points to change a []types.bytes in place, then I guess we need a different implementation for this trim function, perhaps we could skip this "vector reuse procedure" to make our implementation easier